swarm/network/simulation/kademlia.go

```diff
@@ -94,3 +96,53 @@ func (s *Simulation) kademlias() (ks map[enode.ID]*network.Kademlia) {
 	}
 	return ks
 }
+
+func (s *Simulation) WaitNetworkHealth() error {
+	// construct the peer pot, so that kademlia health can be checked
+	addrs := make([][]byte, len(s.Net.Nodes))
+	ids := make([]enode.ID, len(s.Net.Nodes))
+	for i, v := range s.Net.Nodes {
+		ids[i] = v.ID()
+		addrs[i] = v.ID().Bytes()
+	}
+	// construct the peer pot, so that kademlia health can be checked
+	ppmap := network.NewPeerPotMap(2, addrs) //yes i'm hardcoding this
+	check := func(ctx context.Context, id enode.ID) (bool, error) {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		default:
+		}
+
+		node := s.Net.GetNode(id)
+		if node == nil {
+			return false, fmt.Errorf("unknown node: %s", id)
+		}
+		client, err := node.Client()
+		if err != nil {
+			return false, fmt.Errorf("error getting node client: %s", err)
+		}
+		healthy := &network.Health{}
+		if err := client.Call(&healthy, "hive_healthy", ppmap[id.String()]); err != nil {
+			return false, fmt.Errorf("error getting node health: %s", err)
+		}
+		log.Debug(fmt.Sprintf("node %4s healthy: got nearest neighbours: %v, know nearest neighbours: %v, saturated: %v\n%v", id, healthy.GotNN, healthy.KnowNN, healthy.Full))
+		return healthy.KnowNN && healthy.GotNN && healthy.Full, nil
```

Comment:
FYI  `Health.Full` is gone in the kademlia refactor series, and as we talked about on the call will have to be replaced by some evaluation of a combination of "saturation," "potential," and "health."

Reply:
i know. once that's done we can amend accordingly